Create ADR 007-signal-unsupportent-content-in-github.md (#645)
GSA/fedramp-automation: documents/adr/007-signal-unsupportent-content-in-github.md

# 7. Signal Unsupported Contents in the Repository

Date: 2024-08-30

## Status

Pending Community Feedback

## Context

> The FedRAMP OSCAL automation team needs to cleanup the repository periodically, and establish a way to signal to the community areas of the repository that are no longer maintened (e.g., legacy proof-of-concept tools, outdated documentation / guidance, etc.).

Reviewer: Non-blocking: some more typos and cleanup.

```suggestion
The FedRAMP Automation Team must maintain organization of its repositories, including deprecation and migration information for data, tools, and documentation as part of its release strategy.
```

### Possible Solutions

> Signal unsupported code / content in the repository by:

Reviewer: ```suggestion
Several possible solutions for identifying deprecated and migrated data, tools, and documentation are below.

```

> 1. marking the content as **DEPRECATED** in a README

Reviewer: ```suggestion
1. Do nothing, leave all data, tools, and documentation as-is
1. Marking the content as **DEPRECATED** in the relevant READMEs
```

2. **OBSOLETE** in a README

>  - deleting **OBSOLETE** content after some period of time or at specified release 

Reviewer: Thanks to the power of `git` and services like GitHub, GitLab, Codeberg, we get this for free.

```suggestion
 - deleting **OBSOLETE** content from future releases and leaving archival copies in an archive branch for posterity
```

> 2. **OBSOLETE** in a README
>  - deleting **OBSOLETE** content from future releases and leaving archival copies in an archive branch for posterity

Reviewer: ```suggestion
1. Leverage [the current release strategy and semantic versioning](https://github.com/GSA/automate.fedramp.gov/blob/f085168bc485cf764a37c148d3a3109507d4695a/content/about/release.md):
  - identify deprecation of relevant data, tools, and documentation in README with an **OBSOLETE** label in the relevant READMEs;
    - make a minor release with deprecation information in those READMEs;
    - make a major release with deprecated content deleted and new data, tools, and documentation released
```

> 3. marking the content as **ARCHIVED** and moving to an “archived” folder

Reviewer: ```suggestion
1. Mark deprecated content by labelling relevant READMEs as **ARCHIVED** and move all relevant content to “archive” folder
```

   
## Consequences

> The above will be applied to the current repository as follows:  

Reviewer: ```suggestion
For the purposes of this ADR, there is an inventory for this repository of different tools, data, documentation, and their future deprecation below.

```

    * FedRAMP to consider future work on replacing them with report of constraints auto-generated by CI or OSCAL-CLI (from Metaschema constraints)  
* **dist/content/rev5/templates**  
  * Update README clarifying intent of FedRAMP “template” and documenting known validation issues with each template.  Where possilbe, FedRAMP will attempt to fix template validation issues to reduce the amount of errors they produce.  
* **documents** folder  
  * Update README sending community to [https://automate.fedramp.gov](https://automate.fedramp.gov) for documentation going forward.  Add statement about PDF guides no longer being maintained and that existing GH issues submitted about the guides will be address via updates to [https://automate.fedramp.gov](https://automate.fedramp.gov)   
    * **documents/rev4/README.md** \- mark as **OBSOLETE** and remove PDF guides
    * **documents/rev5/README.md** \- mark as **OBSOLETE** and remove PDF guides
* presentations folder  
  * **presentations/data-bites**  
    * Update with latest Data Bites slides  
  * Add **presentations/oscal-implementers**

>     * Update with latest OSCAL Implementers slides

Reviewer: Why not move to the automate.fedramp.gov site? These artifacts balloon git history and are really for documentation and communication, they are not code artifacts. Thoughts from the others?

Author: Good point.  I'll update the ADR accordingly

* **src/examples/README.md** \- **ARCHIVED**

>   * Move all current contents to archived subfolder (e.g., **src/examples/archived** )

Reviewer: With git we can delete and still keep history, no?

```suggestion
  * Delete all current contents
```

* **src/utils/extensions-preview.xsl** \- **ARCHIVED**

>   * Move all current contents to archive subfolder (e.g., **src/examples/archived** )

Reviewer: With git we can delete and still keep history, no?

```suggestion
  * Delete all current contents
```

* **validations** folder \- **ARCHIVED**

>   * Move all current contents to archive subfolder (e.g., **src/validations/archived** )

Reviewer: With git we can delete and still keep history, no?

```suggestion
  * Delete all current contents
```

# 7. Signal Unsupported Contents in the Repository

> Date: 2024-08-30

Reviewer: Important to change the date accordingly.

```suggestion
Date: 2024-09-17
```

## Status

> Pending Community Feedback

Reviewer: Ditto, mark it accepted as no feedback has been received by community over two meetings.

```suggestion
Accepted
```

## Context

The FedRAMP Automation Team must maintain organization of its repositories, including deprecation and migration information for data, tools, and documentation as part of its release strategy.

### Possible Solutions

Several possible solutions for identifying deprecated and migrated data, tools, and documentation are below.

1. Do nothing, leave all data, tools, and documentation as-is
1. Marking the content as **DEPRECATED** in the relevant READMEs
1. Leverage [the current release strategy and semantic versioning](https://github.com/GSA/automate.fedramp.gov/blob/f085168bc485cf764a37c148d3a3109507d4695a/content/about/release.md):
  - identify deprecation of relevant data, tools, and documentation in README with an **OBSOLETE** label in the relevant READMEs;
    - make a minor release with deprecation information in those READMEs;
    - make a major release with deprecated content deleted and new data, tools, and documentation released
1. Mark deprecated content by labelling relevant READMEs as **ARCHIVED** and move all relevant content to “archive” folder

>    

Reviewer: It is time to make a decision, so we shall picture Solution 3.

```suggestion
 ## Decision
 
 The FedRAMP Automation Team will implement Solution 3: identify tools, data, and content for migration, deprecation, remove content accordingly by saving them in an archival minor release and deleting them in a major version release.
   
```

## Consequences

For the purposes of this ADR, there is an inventory for this repository of different tools, data, documentation, and their future deprecation below.

* **dist/content/rev4**  
  * Add README \- FR has transitioned to Rev 5\.  Rev 4 baselines, resources, and templates are no longer maintained but will be kept frozen for historical purposes.
  * Mark as **DEPRECATED**
* **dist/content/rev5/resources**  
  * Archive README \- letting community know FedRAMP is transitioning from *fedramp\_extensions.xml* & *fedramp\_values.json*
    * FedRAMP to consider future work on replacing them with report of constraints auto-generated by CI or OSCAL-CLI (from Metaschema constraints)  
* **dist/content/rev5/templates**  

>   * Update README clarifying intent of FedRAMP “template” and documenting known validation issues with each template.  Where possilbe, FedRAMP will attempt to fix template validation issues to reduce the amount of errors they produce.  

Reviewer: Non-blocking question: what is the intent?

Author: Community often has questions about errors when trying to validate the templates.   I've been able to find and fix a lot of the SSP template validation failures using our new tooling, but there's a handful of errors that remain (some seem like false positives).  I'm thinking we should document those remaining errors somewhere, so the community knows these are known issues.

Reviewer: Per conversation with you and Dave on Friday, we can put a deprecation warning and defer deletion as necessary, it doesn't necessarily have to be immediately. I will close this out for now.

* **documents** folder  
  * Update README sending community to [https://automate.fedramp.gov](https://automate.fedramp.gov) for documentation going forward.  Add statement about PDF guides no longer being maintained and that existing GH issues submitted about the guides will be address via updates to [https://automate.fedramp.gov](https://automate.fedramp.gov)   
    * **documents/rev4/README.md** \- mark as **OBSOLETE** and remove PDF guides
    * **documents/rev5/README.md** \- mark as **OBSOLETE** and remove PDF guides
* presentations folder  
  * **presentations/data-bites**  

>     * Move to automte.fedramp.gov and update with latest Data Bites slides  

Reviewer: Minor, non-blocking: the domain is misspelled.

```suggestion
    * Move to [https://automate.fedramp.gov](https://automate.fedramp.gov) and update with latest Data Bites slides  
```
